File: C/Lab5/Dictionary.c
// Samer Baslan
// sbaslan
// CMPS 12M
// 08-06-2017
// Dictionary.c


#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"


// private types --------------------------------------------------------------

// NodeObj
typedef struct NodeObj{
   char* key;
   char* value;
   struct NodeObj* next;
} NodeObj;

// Node
typedef NodeObj* Node;

//create dictionary struct
typedef struct DictionaryObj{
	Node head;
	Node tail;
	int numItems;
} DictionaryObj;


	

// newNode()
// constructor of the Node type
Node newNode(char* k, char* v){
	Node N = malloc(sizeof(NodeObj));
	assert(N != NULL);
	N-> key = k;
	N-> value = v;
	N-> next = NULL;
	return (N);
}

// freeNode()
// destructor for the Node type
void freeNode(Node* pN){
   if( pN!=NULL && *pN!=NULL ){
      free(*pN);
      *pN = NULL;
   }
}

// public functions


// newDictionary
// Constructor of the Dictionary type
Dictionary newDictionary(void){
	Dictionary D = malloc(sizeof(DictionaryObj));
	assert(D != NULL);
	D->head = NULL;
	D->tail = NULL;
	D->numItems = 0;
	return D;
}


// freeDictionary()
// destructor for the Stack type
void freeDictionary(Dictionary* pD){
   if( pD!=NULL && *pD!=NULL ){
      if( !isEmpty(*pD) ) makeEmpty(*pD);
      free(*pD);
      *pD = NULL;
   }
}

// isEmpty()
// returns 1 (true) if D is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D){
   if( D==NULL ){
      fprintf(stderr, 
              "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   if(D->numItems> 0){
	return 0;
   }else{
	return 1;
   }
}

// size()
// returns the number of (key, value) pairs in D
// pre: none
int size(Dictionary D){
	return D->numItems;
}

// lookup()
// returns the value v such that (k,v) is in D, or returns NULL if no such value exists
// pre: !isEmpty(S)
char* lookup(Dictionary D, char* k){
   Node N = D->head;
   if( D==NULL ){
      fprintf(stderr, "Dictionary Error: calling lookup() on NULL Dictionary\n");
      exit(EXIT_FAILURE);
   }
   if( D->numItems==0 ){
      fprintf(stderr, "Dictionary Error: calling lookup() on empty Dictionary\n");
      exit(EXIT_FAILURE);
   }
   while(N != NULL){
	if(strcmp(N->key, k) == 0)
	return N->value;
	N = N->next;
   }
	return NULL;
}


// insert()
// interts new (key, value) pair into D
// pre: lookup(D, k) == NULL
void insert(Dictionary D, char* k, char* v){
   if(D->numItems == 0){
	D-> head = D-> tail = newNode(k, v);
   }else{
	Node N;
	N = newNode(k, v);
	D->tail -> next = N;
	D-> tail = N;
	}
   D->numItems++;
}

// delete()
// deletes pair with the key k
// pre: lookup(D, k)!= NULL
void delete(Dictionary D, char* k){
   Node N = D->head;
   if(lookup(D, k) == NULL){
	fprintf(stderr, "Dictionary error: Key Not Found\n");
	exit(EXIT_FAILURE);
   }
   if(strcmp(N->key, k) == 0){
	Node P = D->head;
	Node S = P->next;
	D->head = S;
	freeNode(&P);
   }
   else{	
        while( N!= NULL && N->next != NULL){
	if(strcmp(N->next->key, k) == 0){
	Node S = N;
	Node C = N->next;
	S->next = C-> next;
	N = S;
	freeNode(&C);
   }
   N = N->next;
}
}
   D->numItems--;
}

// makeEmpty()
// re-sets D to the empty state
// pre: none

void makeEmpty(Dictionary D){
   D->head = NULL;
   D->tail = NULL;
   D->numItems = 0;
}

// printDictionary()
// pre: none
// prints a text representation of D to the file pointed to by out

void printDictionary(FILE* out, Dictionary D){
   Node N;
   if(D == NULL){
	fprintf(stderr, "Dictionary Error: calling printDictionary() on NULL Dictionary reference\n");
	exit(EXIT_FAILURE);
   }
   for(N=D->head; N != NULL; N=N->next){
	fprintf(out, "%s %s ", N->key, N->value);
	fprintf(out, "\n");
}
}

File: Java/Lab7/Dictionary.java
//Samer Baslan
//sbaslan
//CMPS 12M
//08-17-2017
//BST implementation of Dictionary.java

public class Dictionary implements DictionaryInterface{

   // inner private class Node
   
   private class Node{
      String key;
      String value;
      Node left;
      Node right;
   
      // Node constructor
      Node(String key, String value){
         this.key = key;
         this.value = value;
         left = null;
         right = null;
      }
   }

   // private fields for dictionary class
   private Node root;
   private int numPairs;

   // findKey()
   private Node findKey(Node R, String k){
      if(R == null || k.compareTo(R.key) == 0){
         return R;
      }
      if(k.compareTo(R.key) < 0){
         return findKey(R.left, k);
      }
      else{
         return findKey(R.right, k);
      }
   }

   // findParent()
   private Node findParent(Node N, Node R){
      Node P = null;
      if(N != R){
         P = R;
         while(P.left != N && P.right != N){
            if(N.key.compareTo(P.key) < 0){
               P = P.left;
            }
            else{
               P = P.right;
            }
         }
      }
      return P;
   }

   // findLeftmost()
   private Node findLeftmost(Node R){
      Node L = R;
      if(L != null) for( ; L.left != null; L = L.left) ;
      return L;
   }

   void printInOrder(Node R){
      if(R!=null){
         printInOrder(R.left);
         System.out.println(R.key + " " +R.value);
         printInOrder(R.right);
      }
   }
   
   // deleteAll()
   void deleteAll(Node N){
      if(N != null){
         deleteAll(N.left);
         deleteAll(N.right);
       }
   }

   // public functions ----------------------------------------------------------
   // Dictionary Constructor
   public Dictionary(){
      root = null;
      numPairs = 0;
   }

   // isEmpty()
   // pre: none
   // returns true if this Dictionary is empty, false otherwise
   public boolean isEmpty(){
      if(numPairs == 0){
         return true;
      }
      else{
         return false;
      }
   }
            
   // size()
   // pre: none
   // returns the number of entries in this Dictionary
   public int size(){
      return numPairs;
   }

   // lookup()
   // pre: none
   // returns value associated key, or null reference if no such key exists
   public String lookup(String key){
      Node N;
      N = findKey(root, key);
      return(N == null ? null : N.value);
   }

   // insert()
   // inserts new (key,value) pair into this Dictionary
   // pre: lookup(key)==null
   public void insert(String key, String value) throws DuplicateKeyException{
      if(lookup(key) != null){
         throw new DuplicateKeyException("Dictionary Error: cannot insert() duplicate key: "+key);
      }
      else{
         Node N = new Node(key, value);
         Node B = null;
         Node A = root;
         while(A != null){
            B = A;
            if(key.compareTo(A.key) < 0){
               A = A.left;
            }
            else{
               A = A.right;
            }
         }
         if(B == null){
            root = N;
         }
         else if(key.compareTo(B.key) < 0){
            B.left = N;
         }
         else{
            B.right = N;
         }
         numPairs++; 
      }
   }
         

   // delete()
   // deletes pair with the given key
   // pre: lookup(key)!=null
   public void delete(String key) throws KeyNotFoundException{
      Node N = findKey(root, key);
      if(lookup(key) == null){
         throw new KeyNotFoundException("Dictionary Error: cannot delete() non-existent key: "+key);
      }
      else{
         if(N.left == null && N.right == null){ // case 1
            if(N == root){
               root = null;
            }
            else{
               Node P = findParent(N, root);
               if(P.right == N){
                  P.right = null;
               }
               else{
                  P.left = null;
               }
            }
         }
         else if(N.right == null){ //case 2 left child only
            if(N == root){
               root = N.left;
            }
            else{
               Node P = findParent(N, root);
               if(P.right == N){
                  P.right = N.left;
               }
               else{
                  P.left = N.left;
               }
            }
         }
         else if(N.left == null){ //case 2 right child only
            if(N == root){
               root = N.right;
            }
            else{
               Node P = findParent(N, root);
               if(P.right == N){
                  P.right = N.right;
               }
               else{
                  P.left = N.right;
               }
            }
         }
         else{                    //case 3: 2 children
            Node S = findLeftmost(N.right);
            N.key = S.key;
            N.value = S.value;
            Node P = findParent(S, N);
            if(P.right == S){
               P.right = S.right;
            }
            else{
               P.left = S.right;
            }
         }
         numPairs--;
      }
   }
 
   // makeEmpty()
   // pre: none
   public void makeEmpty(){
      deleteAll(root);
      root = null;
      numPairs = 0;
   }
      
   // toString()
   // returns a String representation of this Dictionary
   // overrides Object's toString() method
   // pre: none
   public String toString(){
      String x = "";
      printInOrder(root);
      return x;
   }
}  
